### ISSUE TITLE
How to do components efficiently?

### ISSUE THREAD

**Issue by serapath** [Matt-Esch/virtual-dom#304]
Probably I'm just missing something obvious, but given I have a component tree like:

```
           |---|
           | R |
           |---|
             |
      |------|------|
      |             |
    |---|         |---|
    | A |         | B |
    |---|         |---|
      |             |
  |---|---|     |---|---| 
  |       |     |       |
|---|   |---| |---|   |---|
| C |   | D | | E |   | F |
|---|   |---| |---|   |---|
```
If my components look like:
```js
function ComponentR (state) {
  return h('div', [state.R,ComponentA(state),ComponentB(state)])
}
function ComponentA (state) {
  return h('div', [ state.A,ComponentC(state),ComponentD(state)])
}
function ComponentB (state) {
  return h('div', [state.B,ComponentE(state),ComponentF(state)])
}
function ComponentC (state) { return h('div', state.C) }
function ComponentD (state) { return h('div', state.D) }
function ComponentE (state) { return h('div', state.E) }
function ComponentF (state) { return h('div', state.F) }
```
I might be able to integrate some kind of event listener or other mechanism into each components to receive updates from user interaction or data updates from a server, but by relying on the `return` statement, I guess I would always need to re-render the whole component-tree

What I imagine is, that given any update, only certain components need to re-render.
* If the update is just affecting a "leaf component", it would just need to re-render itself.
* If the update is affecting the root - the tree might be re-rendered, but some components between the "root" and the "leafs" can eventually be skipped, sometimes even whole component sub trees.

I guess, when it comes to react, methods like `shouldComponentUpdate` and the like are related to that "problem"

**Questions:**
* What should I read to learn more about it?
* Is `vdom-thunk` related to that problem? (If so, could someone point me to an example?)
* Should I change the `return` logic and instead have something like "vdom manipulation", where the whole thing starts of making a "copy" of the current vdom, applying some updates and then updating the dom?

**Comment by Zolmeister** [comment]
The answer is to use [thunks](https://github.com/Matt-Esch/virtual-dom/blob/master/docs/thunk.md)
Relevant issue: https://github.com/Matt-Esch/virtual-dom/issues/303

**Comment by serapath** [comment]
thank you :-)

**Comment by serapath** [comment]
I was reading a bit about those `vdom-thunk`s, but it's still unclear to me how I would apply them to implement the idea i try to sketch below.  
Given the component structure from above, I would like to have a "render mechanism" that works like follows:
```js
// NOTATION (for current state of vtree) - EXAMPLE:
// Tree[Version](Subtree1[Version](...),LeafComponent2[version],Subtree3[version](...),...)
//
// NOTATION (for a user or system induced update through interaction with a component)
// Component.called_by_user_or_system_with_random_update()

// EXAMPLE
 var e     = require("engine-vtree2dom")
 var opts  = {/* some opts for R */}
 var R     = require("component-r")(opts) // R's dependency structure => (see above posts)
 //R.push("R[0]( A[0](C[0],D[0]), B[0](E[0],F[0]) )")
 R.pipe(e)
 //e.vtree === "R[0]( A[0](C[0],D[0]), B[0](E[0],F[0]) )"

// SIMULATION (of interaction)
// for demonstration purposes I assume access to a reference to each component
// used by R or their child components

 A.called_by_user_or_system_with_random_update()
 //A.push("      A[1](C[0],D[0])                   ")//.pipe(R)
 //R.push("R[0]( A[1](C[0],D[0]), B[0](E[0],F[0]) )")//.pipe(e)
 //e.vtree === "R[0]( A[1](C[0],D[0]), B[0](E[0],F[0]) )"
 
 D.called_by_user_or_system_with_random_update()
 //D.push("                D[1]                    ")//.pipe(A)
 //A.push("      A[1](C[0],D[1])                   ")//.pipe(R)
 F.called_by_user_or_system_with_random_update()
 //F.push("                                 F[1]   ")//.pipe(B)
 //B.push("                       B[0](E[0],F[1])  ")//.pipe(R)
 //R.push("R[0]( A[1](C[0],D[0]), B[0](E[0],F[1]) )")//.pipe(e)
 //e.vtree === "R[0]( A[1](C[0],D[0]), B[0](E[0],F[1]) )"
 //R.push("R[0]( A[1](C[0],D[1]), B[0](E[0],F[1]) )")//.pipe(e)
 //e.vtree === "R[0]( A[1](C[0],D[1]), B[0](E[0],F[1]) )"
```

I hope this makes sense and I'm not totally on the wrong track. The above is more pseudo code and a pseudo process that i tried to sketch.

One question I have in mind is:
* `given:` components are implemented
* `given:` engine-vtree2dom is implemented and whenever it receives a new `vtree`, it diffs with the old one and applies the patches to the dom

How would I "mark stuff" in my `vtree` with this `vdom-thunk` thing during each processing step in the component hierarchy, which makes up my **"vtree stream pipeline"**, so that i minimize the cost for creating patches, once the whole built-up `vtree` arrives at the `engine-vtree2dom`?

I would appreciate general feedback about the approach too :-)

**Comment by Zolmeister** [comment]
Your example is hard to follow
We do it like this
Root -> [Child1, Child2]

On change, invalidate thunk cache
Child1.markThunkInvalid() -> Root.markThunkInvalid() -> re-render page

render page -> Root.render() -> [Child1.render(), Child2.thunk (no change)]

Code: https://github.com/Zorium/zorium/blob/master/src/z.coffee

**Comment by serapath** [comment]
I updated my posting - maybe it's now a bit more clear.

#### In your description:
* you executed `render.page()`,
* which executes `Root.render()`,
* which executes `[Child1.render(), Child2.thunk (no change)]`
* and so on ....
**can you tell more about how `render.page()` gets triggered in the first place?**

#### In my example the flow is reversed (I will use the flow described above):
```js
/*
                 |-----|                  |------------|
                 | R-0 |                  | NAME-STATE |
                 |-----|                  |------------|
                    |
           |--------|--------|
           |                 |
         |-----|           |-----|
         | A-0 |           | B-0 |
         |-----|           |-----|
            |                 |
       |----|----|       |----|----|
       |         |       |         |
    |-----|   |-----| |-----|   |-----|
    | C-0 |   | D-0 | | E-0 |   | F-0 |
    |-----|   |-----| |-----|   |-----|
*/
```
* **"ChildA"** gets updated (e.g. through the user clicking a button that is part of ChildA),
* which executes `"ChildA.render()"` and `A.pipes(R)` the resulting `NewVtree` to its "parent",
* A -> which executes `"Root.render()"` and `R.pipes(engine)` the resulting `NewVtree` to the "engine",
* R -> which executes `engine.patch(dom, NewVtree)`,

... my example goes on ...

* **"ChildD"** gets updated by the user or system,
* which executes `ChildD.render()` and `D.pipes(A)` the resulting `NewVtree` to its "parent",
* D -> which executes `ChildA.render()` and `A.pipes(R)` the resulting `NewVtree` to its "parent",
* ... another update somewhere else somehow manages to "squeeze in" ...
* **"ChildF"** gets updated by the user or system,
* which executes `ChildF.render()` and `F.pipes(B)` the resulting `NewVtree` to its "parent",
* F -> which executes `ChildB.render()` and `B.pipes(R)` the resulting `NewVtree` to its "parent",
* B -> which executes `Root.render()` and `R.pipes(engine)` the resulting `NewVtree` to the "engine",
* R -> which executes `engine.patch(dom, NewVtree)`,
* ... the other update paths continues and finishes ...
* A -> which executes `Root.render()` and `R.pipes(e)` the resulting `NewVtree` to the "engine",
* R -> which executes `engine.patch(dom, NewVtree)`,

**Comment by Zolmeister** [comment]
So because thunks are lazy, you never actually call `child.render()` yourself (except within the thunk)
What happens is an upwards chain of 'dirty' flags, which only call `child.render()` during the `engine.patch` event (via thunk)

This line checks thunk dirtyness: https://github.com/Zorium/zorium/blob/master/src/z.coffee#L60
And this line dirties the thunk: https://github.com/Zorium/zorium/blob/master/src/z.coffee#L141

And when the Root node thunk becomes dirty (which should be any change in the tree because it bubbles up) re-render: https://github.com/Zorium/zorium/blob/master/src/router.coffee#L139

**Comment by serapath** [comment]
The main thought behind all of this is, that the need for a `diff` feels wrong in the first place.
1. Because components `return` virtual dom, an ancestor has to render them
2. This means, the subtree has to be traversed looking for more "dirtyness"

If there was something passed into each component, that can be used by the component to push updates, the need for a re-render of the component by a parent would vanish.
In the `levelup` ecosystem, you have a module called `sublevel`, which exposes the same `API` as a regular `db` object, but all **keys** in `.get(..)` or `.put(...)` or other operations are prefixed.

```js
var ComponentA = require('component-a')
var ComponentB = require('component-b')

module.exports = function ComponentR (engine, db) {

  ComponentA(engine.sub('component-a'), db.sub('component-a'))
  ComponentB(engine.sub('component-b'), db.sub('component-b'))

  var containerA = h('div#' + engine.getID() + 'componentA')
  var containerB = h('div#' + engine.getID() + 'componentB')

  var vtree = h('div', [db.get('component-r'), containerA, containerB])
  engine.push(vtree) // updates ComponentR's container with new vtree of ComponentR

}
```
When `ComponentA` or `ComponentB` do an `engine.push(vtree)` this will update the content of `containerA` or `containerB`. Something like that is the general idea, that i'm trying to get at.

If, for example, `ComponentR` decides to replace the `containerA` with something else, then `ComponentA` would not be needed any longer.

Wouldn't something like that make diffing unnecessary?

**Comment by Zolmeister** [comment]
Agreed, there should be a way to do partial diffs on subtrees.
I think the interface though should be a function passed to the component, perhaps through a hook.

```coffee
class Component
  hook: (render) =>
    @state.subscribe =>
      render @render()
  render: ->
    h 'div', 'xxx'
```

My current workaround has changed since before, and is now a bit simpler.
TL;DR - onMount -> listen, onChange -> mark 'dirty', tell parent your dirty (propogate upwards) -> final parent diffs
(note: using streams all the way through was too slow)
https://github.com/Zorium/zorium/blob/master/src/z_thunk.coffee#L26

**Comment by serapath** [comment]
I'm not sure what you mean with `partial diffs on subtrees`.
The code example is hard to read for me too. I'm not used to using coffeescript and classes and es6 so heavily. I'm also not sure why and/or what the hook would do.

I imagine that inside...
```js
function tracker (db, query) {
  return require('level-tracker')(db).trackable().track({ gte: query, lte: query, old: false})
}

module.exports = function ComponentR (engine, db) {

  // ...

  tracker(db, 'component-r/items').on('data', function (items_of_componentR) {
    engine.push(
      h('div', [
        items_of_componentR,
        containerA,
        containerB
      ])
    )
  })

  // ...

}
```

So after the ComponentR was initialized (see code snippet previous comment), the only thing that stays active are those kind of listeners or "piped streams" that a component sets up when initialized.  
(See example above)

It basically says, that whenever an entry in the leveldb (=wrapper around indexDB) changes, the `ComponentR` is being informed and re-renders itself...
* it is not rendered by a parent or ancestor and the parents and ancestors are not re-rendered
* it does not re-render it's children, because if children need re-rendering, they can listen to changes in the leveldb and re-render themselves the same way `ComponentR` does

....that's the whole thought behind it.

---

I'm actually too much of a beginner to really understand virtual-dom's internals and I'm not sure if all the things necessary to make the described code work are in place - if not - support would be great, because then I can try to implement it myself when i have time.

I'm thinking already for some time about combining streams with virtual dom.  
I feel a lot of people see Rxjs as too much of a learning curve to get into the functional standard vocabulary it seems to offer and instead prefer streams, where you make up your custom vocabulary over time. ...at least that's my current perception :-)
...anyway, i'd love to see how virtual dom would work with streams and how that would compare to all the Rxjs stuff I see popping up in lots of places :-)

**Comment by Zolmeister** [comment]
@serapath I tried to clarify the idea in my proposal (in js). Hopefully you also see how streams are used (as state).
https://github.com/Matt-Esch/virtual-dom/issues/343

Having a 'tracker' or any database layer to manage state I think is going a bit too far. I like the idea of components themselves managing their updates, and keeping the interface as simple and pure as possible.

**Comment by serapath** [comment]
What I like is, that levelup can be backed by a `memdb` to just have it in memory but it could also use `indexdb` in the browser or it could replicate back to a server-side leveldb.
There could also be some kind of "bridge" that translates any kind of API into populating a client side leveldb and/or keeping the client side leveldb up-to-date

The levelup ecosystem is diverse and starting from a pure in-memory-thing it can be updated with whatever features are necessary for any given "app requirements" :-)

In the end - if I use something else, for any serious project, I have to sync with a database anyway, so making this more structured feels pragmatic to me.

**Comment by Zolmeister** [comment]
@serapath seems outside the scope of virtual-dom.

**Comment by serapath** [comment]
yes, it's out-of-scope. sry :-)

But your example and mine included some kind of "subscribe to state" and updates trigger a "self-rerender". I agree the state thing, that offers a subscribe could be multiple things, doesnt need to be a levelup thing
